source/opt/function.h

```diff
@@ -262,6 +274,36 @@ inline void Function::AddNonSemanticInstruction(
   non_semantic_.emplace_back(std::move(non_semantic));
 }
 
+template <class It>
+void Function::ReorderBasicBlocks(It begin, It end) {
+  // Asserts to make sure every node in the function is in new_order.
+  assert(ContainsAllBlocksInTheFunction(begin, end));
+
+  // We have a pointer to all the elements in order, so we can release all
+  // pointers in |block_|, and then create the new unique pointers from |{begin,
+  // end}|.
+  std::for_each(blocks_.begin(), blocks_.end(),
+                [](std::unique_ptr<BasicBlock>& bb) { bb.release(); });
+  std::transform(begin, end, blocks_.begin(), [](BasicBlock* bb) {
+    return std::unique_ptr<BasicBlock>(bb);
+  });
+}
+
+template <class It>
+bool Function::ContainsAllBlocksInTheFunction(It begin, It end) {
+  std::vector<BasicBlock*> range(begin, end);
+
+  if (range.size() != blocks_.size()) {
+    return false;
+  }
+
+  std::sort(range.begin(), range.end());
+  for (auto& bb : blocks_) {
+    if (!std::binary_search(range.begin(), range.end(), bb.get())) return false;
+  }
+  return true;
+}
```

Comment:
I think `ContainsAllBlocksInTheFunction` can be just
```
std::unordered_set<BasicBlock*> blocks(begin, end);
std::unordered_set<BasicBlock*> function_blocks;
for (auto&& bb : blocks_) {
    function_blocks.insert(bb.get());
}
return blocks == function_blocks;

Reply:
That would mostly work.  You're missing the case where the input has duplicates: `{BB1, BB2, BB2}`, while the current function is `{BB1,BB2}`.  The keeping the original size check can fix that, which you may have intended.

I chose to use a vector with binary search because in practice vectors are faster, and I usually expect the number of basic blocks to be small.  I did not measure what is actually faster on a large set of shaders.  I can change it.  The runtime is not too important since it is only called for an assert,

Reply:
It's not a big deal as it's only for an assert, as you mention. Up to you!

However, I would probably change the comment to the function from 
> The range must contain every basic block in the function, and no extras.

to

> The range must contain every basic block in the function, and no extras or duplicates.

Just to be a tad more explicit.